README.md

```diff
@@ -1,2 +1,176 @@
+
+... because OAuth2 is an extensible and flexible **framework**. Fosite let's you register new response types, new grant
+types and new response key value pares. This is useful, if you want to provide OpenID Connect on top of your
+OAuth2 stack. Or custom assertions, what ever you like and as long as it is secure. ;)
+
+### Tokens
+
+Tokens are formatted as `<key>.<signature>`. This is beneficial if you want to keep tokens encrypted at rest.
+To validate a token in a OAuth2 grant, you could first check if the key matches the signature and then lookup the signature
+in your persistent storage (e.g. MySQL). If your persistent storage is intruded (e.g. by SQL injection), an attacker would
+only have access to the token signatures and would be, because he does not know the key, unable to use them for authorization.
+
+A token generated by `generator.CryptoGenerator` looks like:
+
+```
+GUULhK6Od/7UAlnKvMau8APHSKXSRwm9aoOk56SHBns.JDJhJDEwJDdwVmpCQmJLYzM2VDg1VHJ5aEdVOE81NVdRSkt6bHBHR1QwOC9pbTNFWmpQRXliTWRPeDQy
+```
+
+## Usage
+
+This section is WIP and we welcome discussions via PRs or in the issues.
+
+### Store
+
+### Authorize Endpoint
+
+```go
+var r *http.Request // we're assuming that we are inside a http.Handler
+var rw http.ResponseWriter  // we're assuming that we are inside a http.Handler
+
+var store fosite.Storage // needs to be implemented or by using some library
+config := fosite.NewDefaultConfig()
+oauth := fosite.NewOAuth(config)
+authorizeRequest, err := oauth.NewAuthorizeRequest(r, store)
+if err != nil {
+    oauth.RedirectError(rw, error)
+    // or, for example: oauth.WriteError(rw, error)
+    return
+}
+
+// you have now access to authorizeRequest.Scope, ...Code ...ResponseTypes ...Scopes ...
+
+// decide what to do based on scope and response type
+// e.g: response, err = oauth.HandleAuthorizeRequest(authorizeRequest)
+
+// set up a session
+// session := oauth2.NewAuthorizeSession(123)
+// session.SetExtra(extra interface{})
+
+// persist that stuff in the database
+// err = oauth2.PersistAuthorizeRequest(authorizeRequest, session) // sets e.g. session.Persistent = true
+
+// finally, persist code in store and send response
+// e.g: oauth2.WriteResponse(rw, response, session)
+```
+
+Because each component returns a different type, we can be (if safeguards are installed) quite sure, that the developer
+implemented the work flow the right way:
+
+1. `NewAuthorizeRequest(args...) *AuthorizeRequest`: Fetch authorize request information
+2. do whatever you like
+3. `HandleAuthorizeRequest(args...) *AuthorizeResponse`: Handle authorize request (check scopes and response types, hydrate response...)
+4. do whatever you like
+5. `oauth2.NewAuthorizeSession(*AuthorizeResponse) *AuthorizeSession`: A session
+6. do whatever you like, e.g. `session.SetExtra(map[string]interface{"foo": "bar"})`
+7. `oauth2.PersistAuthorizeRequest` persists the request in the database so the token endpoint can look up information
```

Comment:
It sort of seems like this should just happen as part of HandleAuthorizeRequest? You could create the session after NewAuthorize request and pass it into Handle.